src/thermo/PengRobinson.cpp

```diff
@@ -0,0 +1,955 @@
+        }
+        m_pp[k] = m_pp[k] * dalphadT_vec_Curr_[k] / alpha_vec_Curr_[k];
+    }
+
+
+    for (size_t k = 0; k < m_kk; k++) {
+        coeff1 = m_b_current * (m_pp[k] + m_tmpV[k]) - daAlphadT * b_vec_Curr_[k];
+        sbar[k] += GasConstant * log(GasConstant * T / (refP * mv))
+                + GasConstant
+                + GasConstant * log(mv / vmb)
+                + GasConstant * b_vec_Curr_[k] / vmb
+                - coeff1* log(vpb2 / vmb2) / den1
+                - b_vec_Curr_[k] * mv * daAlphadT / den2 / m_b_current;
+    }
+    pressureDerivatives();
+    getPartialMolarVolumes(m_partialMolarVolumes.data());
+    for (size_t k = 0; k < m_kk; k++) {
+        sbar[k] -= m_partialMolarVolumes[k] * dpdT_;
+    }
+}
+
+void PengRobinson::getPartialMolarIntEnergies(double* ubar) const
+{
+    getIntEnergy_RT(ubar);
+    scale(ubar, ubar+m_kk, ubar, RT());
+}
+
+void PengRobinson::getPartialMolarCp(double* cpbar) const
+{
+    getCp_R(cpbar);
+    scale(cpbar, cpbar+m_kk, cpbar, GasConstant);
+}
+
+void PengRobinson::getPartialMolarVolumes(double* vbar) const
+{
+    for (size_t k = 0; k < m_kk; k++) {
+        m_pp[k] = 0.0;
+        for (size_t i = 0; i < m_kk; i++) {
+            size_t counter = k + m_kk*i;
+            m_pp[k] += moleFractions_[i] * aAlpha_vec_Curr_[counter];
+        }
+    }
+
+    double mv = molarVolume();
+    double vmb = mv - m_b_current;
+    double vpb = mv + m_b_current;
+    double fac = mv * mv + 2 * mv * m_b_current - m_b_current * m_b_current;
+    double fac2 = fac * fac;
+    double RTkelvin = RT(); 
+
+    for (size_t k = 0; k < m_kk; k++) {
+        double num = (RTkelvin + RTkelvin * m_b_current/ vmb + RTkelvin * b_vec_Curr_[k] / vmb
+                          + RTkelvin * m_b_current * b_vec_Curr_[k] /(vmb * vmb)
+                          - 2 * mv * m_pp[k] / fac
+                          + 2 * mv * vmb * m_aAlpha_current * b_vec_Curr_[k] / fac2
+                         );
+        double denom = (pressure() + RTkelvin * m_b_current / (vmb * vmb)
+                            + m_aAlpha_current/fac
+                            - 2 * mv* vpb *m_aAlpha_current / fac2
+                         );
+        vbar[k] = num / denom;
+    }
+}
+
+double PengRobinson::speciesCritTemperature(double a, double b) const
+{
+    double pc, tc, vc;
+    calcCriticalConditions(a, b, pc, tc, vc);
+    return tc;
+}
+
+double PengRobinson::critTemperature() const
+{
+    double pc, tc, vc;
+    calcCriticalConditions(m_a_current, m_b_current, pc, tc, vc);
+    return tc;
+}
+
+double PengRobinson::critPressure() const
+{
+    double pc, tc, vc;
+    calcCriticalConditions(m_a_current, m_b_current, pc, tc, vc);
+    return pc;
+}
+
+double PengRobinson::critVolume() const
+{
+    double pc, tc, vc;
+    calcCriticalConditions(m_a_current, m_b_current, pc, tc, vc);
+    return vc;
+}
```

Comment:
If a virtual `calcCriticalConditions` method were declared on the `MixtureFugacityTP` class, most of these definitions could be moved to that class as well.

Reply:
There is no problem to move these functions to the parent class; the pengRobinson will work fine.  However all of these functions are defined separately in RedlichKwongMFTP class, where additional temperature dependent parameters are also passed as arguments to calcCriticalConditions(). (I am not sure, why that is actually needed.) In any case, multiple copies of these functions will be defined, either in both child classes (R-K and P-R) or a parent class and R-K class.

Reply:
Corrected. A virtual `calcCriticalConditions `function is defined in `MixtureFugacityTP `class. The calculation of critical properties (i.e. `critTemperature`, `critPressure`, `critVolume`, `critDensity `and `critCompressibility`) has been moved to the `MixtureFugacityTP ` class as well.